sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,260 @@
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the passed in
+        /// <see cref="ReadOnlyMemory{Byte}"/>.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the passed in
+        /// byte array.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromBytes(byte[] data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance using the passed in string.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromString(string data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The passed in stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromStream(Stream stream) =>
+            FromStreamAsync(stream, false).EnsureCompleted();
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The passed in stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="cancellationToken">An optional<see cref="CancellationToken"/> instance to signal
+        /// the request to cancel the operation.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default) =>
+            await FromStreamAsync(stream, true, cancellationToken).ConfigureAwait(false);
+
+        private static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+            if (stream.CanSeek && stream.Length > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(stream),
+                    "Stream length must be less than Int32.MaxValue");
+            }
+            using var memoryStream = new MemoryStream();
```

Comment:
Consider special-casing MemoryStream as an input:
```C#
MemoryStream inputMemoryStream = stream as MemoryStream;

if (inputMemoryStream?.TryGetBuffer(out ArraySegment<byte> buffer) ?? false)
{
    return new BinaryData(new ReadOnlyMemory<byte>(buffer));
}

using (MemoryStream memoryStream = hasLength ? new MemoryStream(length) : new MemoryStream())
{
    ...
...

(Also merged in what I think Krzysztof's comment implies)

Reply:
Why use TryGetBuffer as opposed to GetBuffer?

Reply:
nvm I guess GetBuffer can only be used when we own the stream end to end.

Reply:
Updated to:

```cs
            MemoryStream? inputMemoryStream = stream as MemoryStream;
            ArraySegment<byte> buffer = default;

            if (inputMemoryStream?.TryGetBuffer(out buffer) ?? false)
            {
                return new BinaryData(buffer.Array.AsMemory(buffer.Offset, buffer.Count));
            }

            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream((int)stream.Length) : new MemoryStream();
            if (async)
            {
                await stream.CopyToAsync(memoryStream, CopyToBufferSize, cancellationToken).ConfigureAwait(false);
            }
            else
            {
                stream.CopyTo(memoryStream);
            }
            return new BinaryData(memoryStream.GetBuffer().AsMemory(0, (int)memoryStream.Position));
```